### DISCUSSION TITLE
Multi Synth/Sound Module patches onto one comboBox

### DISCUSSION THREAD

**Discussion by jfk-france** [RomanKubiak/ctrlr#694] [Q&A]
Bonjour everybody,
I am new here, this is a first. I don't have much experience in developing nor in creating plugins. However, I'd like to learn.
I was initially using the old site (ctrl.org - ver5.3.198) a while back, then had to move on to other things. Now I have time again. With that said, I have been able to get a few things running (reading and using lots of the examples within here) but not to any of your level.

Here is what I'd like to achieve (something similar to what - SoundDiver did), well the basic. I have a whole bunch of old school sound modules (which I still liked even if they are cheesy). I'd love to get all those synth/sound module's Patches name (hardcoded right now) into a couple of dropdown  by category (guitar, synth, beats, etc) regardless of the synth/sound module.

As an example, in the Guitar dropdown, I could have 10 Patches from Synth 1, 20 patches from synth 2, etc..

This might be relevant to how I design this:
1. Version = 5.6.30, Build date = SAT, 16 MAR 2024, 02:00:00 EST, Branch = master, Juce = 6.0.8, libusb = 1.0.19, liblo = 0.28, lua = Lua 5.1.4, luabind = 0.9.0, boost = 1.57.0,
2. I am on an Intel MAC OSX 13.7. Renamed all my midiDevices with specific names. (Orbit, XtremeLead, XV-5080, NL2, etc...
3. 8 UNITOR/AMT8 Midi Devices - meaning 64 MIDI Device available to be read in the panel.

Here are my questions on the below code:
1. I am getting an error (see below notes) - I read many post on Ctrlr and CrtlrX, still unable to figure it out.
2. I would like to customize the ComboBox so that it displays items in multiple columns if there are more than 25 items in the ComboBox list, instead of filling a single column and stretching the popup menu beyond the size of the UI to the edges of the screen. Similarly to JUCE

----CODE START HERE-----
SendBassBankPatchChange = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)
    if panel:getBootstrapState() then
        return
    end

-- get Current MIDI information
    local deviceMidiOutput = panel:getProperty("panelMidiOutputDevice")
    local deviceMidichannelOut = panel:getProperty("panelMidiOutputChannelDevice")
    local deviceMidiInput = panel:getProperty("panelMidiInputDevice")
    local deviceMidichannelIn = panel:getProperty("panelMidiInputChannelDevice")

--get info the Combobox
    if value == 0 then
        deviceMidiOut = "Orbit"
        deviceMidichannelOut = 1
        bankValue = 2
        patchValue = 14
    elseif value == 1 then
        deviceMidiOut = "XtremeLead"
        deviceMidichannelOut = 1
        bankValue = 0
        patchValue = 25
... to the entire list (some have 200 patches)
    end

--updating the MIDI Info with the If statement above var
    panel:setPropertyString("panelMidiOutputDevice", deviceMidiOut) -- set the proper Out Device based on the patch 
    panel:setProperty("panelMidiOutputChannelDevice", deviceMidichannelOut, true) -- set the proper Channel from Patch
    --WORKING

--Send BankChange
    local statusByte = 0xB0 + (deviceMidichannelOut - 1) -- use this code for dynamic channel changes
    local midinumber = 32 -- WORKING

    --tried this - NOT WORKING, per different posts in ctrl and ctrlx GitHub
        --local mod = panel:getModulatorByName("modulator-1")
        --local midinumber = tonumber(mod:getMidiMessage():getProperty("midiMessageCtrlrNumber"))

        --valueGroupName = panel:getModulatorByName("modulator-1"):getComponent():getProperty("midiMessageCtrlrNumber")
        --console("Displays valueGroupName : "..(valueGroupName))

        --valueCustomNameGroup = mod:getProperty("midiMessageCtrlrNumber")
        --console("Displays valueCustomNameGroup : "..(valueCustomNameGroup))

    local bankChange = CtrlrMidiMessage(MemoryBlock({statusByte, midinumber, bankValue}))
    panel:sendMidiMessageNow(bankChange)

--send Patch change
    statusByte = 0xC0 + (deviceMidichannelOut - 1) -- use this code for dynamic channel changes
    local programChange = CtrlrMidiMessage(MemoryBlock({statusByte, patchValue}))
    panel:sendMidiMessageNow(programChange)

end

Merci beaucoup - Thanks a million for any help.

**Comment by dnaldoog** [comment]
Hi there @jfk-france 

I would not use _uiCombo_ in this case. It is better to use [**JUCE** _PopupMenu_()](https://docs.juce.com/master/classPopupMenu.html)

Here is a demo panel (attached) by Atom that should get you in the right direction. It includes a more complex panel I made for the JD-990.

Also [this](https://sourceforge.net/projects/jv-1010-editor/files/linux/) panel for the Roland JV-1010 has a complex PopupMenu sttructure.

Don't hesitate to ask any questions as _PopupMenu_() can be difficult.

The main thing to aim for in nested menu items is that each selection must return a unique number.

<pre>
	local ret = m:show(0,0,0,0)

	if ret ~= 0 then
		sendMidi(t[ret]) -- where t{} might be a lookup table with various CC program change, sysex commands etc
	end

t={
"F0 00 02 01 F7",
"F0 00 0A 03 F7",
"B0 07 7F",
}
</pre>

[DEMO.zip](https://github.com/user-attachments/files/17896010/DEMO.zip)

**Reply by jfk-france** [reply]
Hi @dnaldoog ,
Thanks for this, This helped tremendously and I got it to work. Unfortunately, I am unable to open your JV-1010, I got a weird XML error, which unfortunately, I cannot reproduce, it simply won't open.

As I am not familiar with "hard" development, I went to the JUCE PopupMenu() as suggested, and found this function: 
void PopupMenu::addColumnBreak(), which I tried to incorporate as sm:addColumnBreak() and unfortunately I get a method disable, so it's probably not within Ctrlr to be used?

Thanks a lot- Merci beaucoup.

**Comment by dnaldoog** [comment]
Hi @jfk-france,

Yes not all JUCE methods are bound to lua in Ctrlr - that method is no doubt one of them.

I just down loaded the JV-1010 and opened it in Ctrlr. It loads ok. You need the file attached, not any of the other files and I am using Ctrlr 5.3.201 in Windows. If on mac I recommend 5.3.198 because I don't think it will open in Damien's CtrlrX as yet.

**Reply by jfk-france** [reply]
Hi @dnaldoog,
I re-downloaded the panel and opened it with your suggested version and I got the error again:
CtrlrPanel::openXmlPanel unknown file type.

Also am getting a whole bunch or "red' param in other panels. I recall something in multi0pl post talking about ctrlrIDs.xml...Which to me is way over my head. I think.

Thanks for your continued support, much appreciated.

**Comment by dnaldoog** [comment]
What version of Ctrlr are you using?

**Reply by jfk-france** [reply]
I have 3 version for MAC:
5.3.198
5.6.23
5.6.30.1

I ran your panel on 5.3.198 as you suggested.

**Comment by dnaldoog** [comment]
I am at a loss to know what is going on because it works for me when I open in a panel on a virtual box macOS Mojave. Maybe it’s the Mac version you’re running?

**Reply by dnaldoog** [reply]
... or maybe you are trying to lad a zipped file? I've done that!

**Reply by jfk-france** [reply]
Hi @dnaldoog ,
I figured it out 1/2 of the battle - I think - Hurray!!!
For info, here is what was happening, probably because I am a newbie. 
1. When trying to open the panel, I already had the Ctrlr app open, and was getting the error mentioned above.
2. I decided to double click on the panel itself, and MAC gave me some security stuff....I allowed it to open and voila.

However if I still go the other route, open Ctrlr first and then open your panel via the app, I still get the same error - strange.

I do have 2 more questions for you:
1. I am unable to go into Panel MODE - Is that by design?
3. I love this panel - I have a Roland XV-5080 fully loaded - this panel could probably work so well for it.

Lastly, I really wanted to thank you for taking precious time out of your schedule to answer my questions, much much appreciate.

**Comment by dnaldoog** [comment]
That's strange - When I use  5.3.198 on VirtualBox and open the .**bpanelz** file in Ctrlr I can easily go into panel mode, lua editor mode etc - there is an _executable_,_AU_ and _VST_ version as well as the .**bpanelz** file - only on the **bpanelz** file can you access the source code.

**Reply by damiensellier** [reply]
Double clicking on panels to open them on mac never worked for me. I have to open via the file menu inside ctrlr.

**Reply by dnaldoog** [reply]
Yes `File->Open` always

**Reply by jfk-france** [reply]
@dnaldoog and @damiensellier ,
Thank you so much for your continued support. I figured it all it: 4 Mac's  later with different OS, different CPU, different CtrlR version, and more importantly: a family getting mad at me for using their "stuff".  downloaded the wrong file from the link provide by @dnaldoog . I downloaded the "exe" vs. the bpanelz ( which I figured out in your last comment).

As I said - I am a newbie....now I am ready to dive into it.

Merci à vous!!! = Thanks to the both of you.!!!
It is truly appreciated. I'll have lot more questions very soon, I will close this conversation as resolved.

If anytime, you need me to test on MAC...I have plenty of those at home,,different OS, different years, different CPU. I would be happy to help.

**Reply by damiensellier** [reply]
Salut @jfk-france ,
Apparement tu as plusieurs mac à disposition. J'aurais peut être un service à te demander pour vérifier un bug que je suis en train de corriger. 

https://github.com/damiensellier/CtrlrX/issues/18#issuecomment-2509206089

**CtrlrX v5.6.30.1 macOS Intel**
[CtrlrX-5.6.30.1-APP-macOS-Intel.zip](https://github.com/RomanKubiak/ctrlr/files/14659063/CtrlrX-5.6.30.1-APP-macOS-Intel.zip)

C'est quand tu exportes un panel en "restricted instance" à partir du File Menu. Avec certaines version, quand tu fais export, la fenêtre mac native qui montre les dossiers et fichiers ne s'affiche pas (du moins sur mon macbook). Avec Mojave et ventura c'est ok mais pas avec catalina. Je cherche à essayer avec d'autres version de macOS pour voir si ca bug ou pas. Tu as quelles versions à dispo?

Merci

Damien

**Comment by jfk-france** [comment]
Salut @damiensellier ,

A dispo, j'ai the following:
1. MacBook Air M2 (2024) - Sequoia - Impossible d'ouvrir "CtrlX v5.6.30.2-APP-MacOS-Silicon.zip"
2. MacBook Pro i7 (2017) - Ventura -la fenêtre MAC native s'ouvre bien (par contre je ne sais pas quoi faire une fois là). Après la fenêtre ouverte, I click sur "cancel" et je reçois une erreur "Panel Export - Failed to export panel as standalone instance. MAC native, browse for directory dialog failed".
3. MacBook Air i5 (2014) - Big Sur - aucune fenêtre - error.
4. MacBook Air i2 (2010) - High Sierra - la fenêtre MAC native s'ouvre bien (par contre je ne sais pas quoi faire une fois là). Après la fenêtre ouverte, I click sur "cancel" et je reçois une erreur "Panel Export - Failed to export panel as standalone instance. MAC native, browse for directory dialog failed".

Malheureusement, je n'ai plus de Catalina sous la main, mais je vais voir avec un ami, si il en a encore un.

Cordialement,
Igor

**Reply by damiensellier** [reply]
Thanks Igor, it really helped! A bientôt ;)

**Reply by jfk-france** [reply]
I'll let you know if I can get my hands on a Catalina. If you need anything else, please don't hesitate to ask.

**Reply by damiensellier** [reply]
That's fine I already tested on Catalina, the one missing in the list is Monterey. I'm going to update the way the file manager is opening with an async request.
N'hésite pas si tu as besoin d'aide pour autre chose. 
Damien

**Comment by jfk-france** [comment]
I'll look for a Montery one - might take me a couple of days to find one, if any.
J'aurais surement plein de questions :-). Merci et a bientôt.
